frontend/device/wakeupmgr.lua

```diff
@@ -0,0 +1,143 @@
+--[[--
+RTC wakeup interface.
+
+Many devices can schedule hardware wakeups with a real time clock alarm.
+On embedded devices this can typically be easily manipulated by the user
+through `/sys/class/rtc/rtc0/wakealarm`. Some, like the Kobo Aura H2O,
+can only schedule wakeups through ioctl.
+
+See @{ffi.rtc} for implementation details.
+
+See also: <https://linux.die.net/man/4/rtc>.
+--]]
+
+local RTC = require("ffi/rtc")
+local logger = require("logger")
+
+--[[--
+WakeupMgr base class.
+
+@table WakeupMgr
+--]]
+local WakeupMgr = {
+    dev_rtc = "/dev/rtc0", -- RTC device
+    _task_queue = {},
+}
+
+--[[--
+Initiate a WakeupMgr instance.
+
+@usage
+local WakeupMgr = require("device/wakeupmgr")
+local wakeup_mgr = WakeupMgr:new{
+    -- The default is `/dev/rtc0`, but some devices have more than one RTC.
+    -- You might therefore need to use `/dev/rtc1`, etc.
+    dev_rtc = "/dev/rtc0",
+}
+--]]
+function WakeupMgr:new(o)
+    o = o or {}
+    setmetatable(o, self)
+    self.__index = self
+    if o.init then o:init() end
+    return o
+end
+
+--[[--
+Add a task to the queue.
+
+@todo Group by type to avoid useless wakeups.
+For example, maintenance, sync, and shutdown.
+I'm not sure if the distinction between maintenance and sync makes sense
+but it's wifi on vs. off.
+--]]
+function WakeupMgr:addTask(seconds_from_now, callback)
+    if not type(seconds_from_now) == "number" and not type(callback) == "function" then return end
+
+    local epoch = RTC:secondsFromNowToEpoch(seconds_from_now)
+
+    local old_upcoming_task = (self._task_queue[1] or {}).epoch
```

Comment:
I initially stumble on that, "task" being assigned an timestamp, while intuitively, a task is a callback.
So, may be just a comment saying "a task here is a combo of (epoch timestamp, callback).

Nothing more to add about WakeupMrg, looks quite fine and explicite.